### ISSUE TITLE
Change request body in pre-request script

### ISSUE THREAD

**Issue by coder-free** [a85/POSTMan-Chrome-Extension#4808]
I want crypt the request body.
For example,
request body set to this:
![image](https://user-images.githubusercontent.com/2630290/42554373-32d707e0-8517-11e8-80c0-64726b232b1b.png)

and pre-request script set to this:
![image](https://user-images.githubusercontent.com/2630290/42554480-a3d8c0a0-8517-11e8-9b39-b64d5989df6e.png)

and logged the "encryptData" value is encrypted. But the server received data is not encrypt, still is '{"a":"b"}'

**Comment by numaanashraf** [comment]
@zfavourite99 We do not support changing request.body through scripts at the moment. A possible workaround for your use case is to use environments, and keep unencrypted request body in the script.

<img width="615" alt="screen shot 2018-07-11 at 12 55 36 pm" src="https://user-images.githubusercontent.com/3295568/42556603-bfc8324a-8509-11e8-941b-53d28b7da116.png">
<img width="536" alt="screen shot 2018-07-11 at 12 55 40 pm" src="https://user-images.githubusercontent.com/3295568/42556606-c1cb6954-8509-11e8-92e9-bcca154e8c8a.png">

**Comment by numaanashraf** [comment]
Related to #4631

**Comment by coder-free** [comment]
I have many request all is this function to encrypt. So, I created a collection for these requests. I hope the encrypt and decrypt scripts in the collection. Like this:
![image](https://user-images.githubusercontent.com/2630290/42558368-926f2d7a-8523-11e8-9cc3-8232013a4303.png)
If use this "possible workaround", then I need modify every request pre-request and test scripts, that is tedious.
So, I hope postman has this feature in the future. Thanks.
And #4809 feature is twinning.

**Comment by feizeikesi** [comment]
+1

**Comment by madmax911** [comment]
> @zfavourite99 We do not support changing request.body through scripts at the moment. A possible workaround for your use case is to use environments, and keep unencrypted request body in the script.
> 
> <img alt="screen shot 2018-07-11 at 12 55 36 pm" width="615" src="https://user-images.githubusercontent.com/3295568/42556603-bfc8324a-8509-11e8-941b-53d28b7da116.png">
> 
> <img alt="screen shot 2018-07-11 at 12 55 40 pm" width="536" src="https://user-images.githubusercontent.com/3295568/42556606-c1cb6954-8509-11e8-92e9-bcca154e8c8a.png">

That gave me an idea to solve my CRLF issue - thanks man!

![image](https://user-images.githubusercontent.com/11963048/49892914-2b10ed80-fe18-11e8-9a1d-8f27733982c7.png)

**Comment by mkulgeorge** [comment]
Hi numaanashraf,

Storing the value in the environment variable and manipulating that in pre-request Script and assigning that environment variable to the body. 
Very good workaround for changing the POST body parameters dynamically.
This was very useful.

Thanks.

**Comment by shamasis** [comment]
We added mutating URLs and headers for requests from within pre-request script ... mutating other parts like auth and body (to some extent) would follow in short to mid term. :-) Meanwhile, in UI we need to have ways to see what exactly was sent (since programmatic changes are hard to spot.)

**Comment by kibalion2427** [comment]
@madmax911 Your solution was very helpfull for me. Thanks

**Comment by micmania1** [comment]
I've just hit this problem too. It would be great if there were hooks for after the environment variables had been set and the request was ready to send. I had to generate a signature based on the request body and while the environment variable solution `{{encrypted_body}}` would have worked, it makes things very hard to follow.

I need to apply a signature to every request in my collection which is why the `{{encrypted_body}}` wasn't my preferred solution.

I ended up taking the body and doing a manual string replace to generate the signature and let the environment variable do its thing later in the postman request cycle.

eg.
```
// The session ID can't be re-used so we need to create one on the fly
const sessionId = 'session-' + Date.now();
pm.environment.set('REQUEST_SESSION', sessionId);

// Hacky but works for my case
// Note: this does not modify the body
const body = pm.request.body.raw.replace(/{{REQUEST_SESSION}}/g, sessionId);

// Do whatever you need to generate the signature
const secret = pm.environment.get('API_SECRET');
const signature = customGenerateSignatureStuff(body, secret);
pm.environment.set('REQUEST_SIGNATURE', signature);
```

In my case, even a function to parse the body and place the environment variables would be handy, even if it didn't mutate the actual body. That would mean I get consistent variable replacement and don't have to use my own hacky string replace. There may be a function which does this, but i'm not too familiar with postman and couldn't find one.

**Comment by banstola** [comment]
Not sure if this is related but I am running into same problem. I need a way to paste a raw http request string like below and generate postman request. 
```
POST /search?q=test HTTP/2
Host: www.bing.com
User-Agent: curl/7.54.0
Accept: */*

{"key":"value"}
```

What I do currently is paste this whole HTTP request as request body and using JS I parse the HTTP request and completely reconstruct the request and send it. It works but it's all too magical as I can not see the request in the UI. Would appreciate if anyone have solved this?

**Comment by StrohmyerTR** [comment]
> We added mutating URLs and headers for requests from within pre-request script ... **mutating** other parts like auth and **body** (to some extent) **would follow in short to mid term**. :-) Meanwhile, in UI we need to have ways to see what exactly was sent (since programmatic changes are hard to spot.)

---------------------------------------------------------------------------------------------------------------

Any updates on this? This comment is from one year ago.

It confuses me as to why pm.request.body is immutable when nearly everything else is mutable in the pm.request object.

Please post an update on this with a timeline and/or a solid "we're not going to implement this". 

Thanks.

**Comment by eborel-yseop** [comment]
Also interested to have a timeline or a definitive answer on this matter. It is most frustrating :slightly_frowning_face:

**Comment by dem** [comment]
It will be quite usable feature when implemented.

Especially for cases when you are dealing with encryption, signature and converting original body.

Hit into this problem today.

**Comment by topsoft-support** [comment]
+1
I need encrypt body  too！

**Comment by JeffLoo-ong** [comment]
We generate an HMAC token to tag along with our request based on the body.... I can't seem to manipulate the body correctly with environment variables to ensure that each request is a specifically tagged user when using the CryptoJS encrypt function on the body to make a key :/ At this rate I believe it's an issue with the way it reads in the Postman body in pre-request extracted version of the body VS when it actually reads the body to make the key since I am seeing two entirely different generated keys :(

I'm the sole supporter of keeping Postman alive at my company and features like this would give me more ammo to push it through!

**Comment by arlemi** [comment]
@JeffLoo-ong This looks like a good question for community.postman.com since this isn't really a bug. More people will be able to help you there. 🙂

**Comment by StrohmyerTR** [comment]
> @JeffLoo-ong This looks like a good question for community.postman.com since this isn't really a bug. More people will be able to help you there. 🙂

**Everything** else in Postman is programmatically dynamic, so it's hardly a feature request. No one here is looking for "help", we know what we want/need, and are asking for this change to be implemented.

**Comment by eborel-yseop** [comment]
> @JeffLoo-ong This looks like a good question for community.postman.com since this isn't really a bug. More people will be able to help you there.

As a matter of fact I already cited this issue in [this thread](https://community.postman.com/t/bug-pm-request-body-update-doesnt-work/12828/4) where no answers was prodived. I think it's a valid request to ask for this behavior to be implemented.

**Comment by xiaomaolver** [comment]
Many developers want this feature

Me too.

**Comment by alimlong** [comment]
need to modify request body via pre-request script

**Comment by newARTixBazon** [comment]
need to modify request body via pre-request script

**Comment by BruceZhang1993** [comment]
the workaround should work for most situations, except if I want to encrypt the whole original body before sending it.

**Comment by ccjeaty** [comment]
> @zfavourite99 We do not support changing request.body through scripts at the moment. A possible workaround for your use case is to use environments, and keep unencrypted request body in the script.
> 
> <img alt="screen shot 2018-07-11 at 12 55 36 pm" width="615" src="https://user-images.githubusercontent.com/3295568/42556603-bfc8324a-8509-11e8-941b-53d28b7da116.png">
> 
> <img alt="screen shot 2018-07-11 at 12 55 40 pm" width="536" src="https://user-images.githubusercontent.com/3295568/42556606-c1cb6954-8509-11e8-92e9-bcca154e8c8a.png">

that all my data will put in `Pre-script` tab!  i think this is not the best solution.  LOL

**Comment by StrohmyerTR** [comment]
> > @zfavourite99 We do not support changing request.body through scripts at the moment. A possible workaround for your use case is to use environments, and keep unencrypted request body in the script.
> > <img alt="screen shot 2018-07-11 at 12 55 36 pm" width="615" src="https://user-images.githubusercontent.com/3295568/42556603-bfc8324a-8509-11e8-941b-53d28b7da116.png">
> > <img alt="screen shot 2018-07-11 at 12 55 40 pm" width="536" src="https://user-images.githubusercontent.com/3295568/42556606-c1cb6954-8509-11e8-92e9-bcca154e8c8a.png">
> 
> that all my data will put in `Pre-script` tab! i think this is not the best solution. LOL

When will this be supported? It's the only facet which cannot be modified programmatically. It's a nuisance.

**Comment by sph3rex** [comment]
Given a 1mb json with let's say 20 keys and a second api URL that returns a 100Kb json data that needs to be used in the initial JSON, do you have a solution for it rather than hardcoding the json in pre-request? This feature is really needed, and looks like you decided to ignore it for 2 years. I'm constantly bumping into it. The "later"

**Comment by Granfalloner** [comment]
Any updates on this feature?

**Comment by arlemi** [comment]
@Granfalloner Not at the moment, we'll update this issue when there is something new.

**Comment by lvl99** [comment]
@arlemi is there anywhere where we can vote up this feature so it can have higher priority?

**Comment by arlemi** [comment]
@lvl99  You can upvote by adding a 👍 on the first comment in this issue.

**Comment by solirpa** [comment]
priority 👍👍👍
our program always needs to change a obj to form-data.
Such as:
```
{
 product: {
 prefix：xxx,
 name: xxx，
 desc: xxx,
 ...
}

# form-data
dproduct: 1,
dproduct_prefix: xxxx,
dproduct_name: xxxx,
dproduct_descx: xxxx,
...
```

**Comment by baohouse** [comment]
My use case is injecting an "extensions" key into a GraphQL query request (yes, I realize that is not mentioned in the GraphQL spec; "extensions" is part of response, but not request). If we want to use Postman's GraphQL editor, then the workaround mentioned above will not allow us to do so. For now, my workaround is running a local Express.js proxy server that transforms request bodies before sending it off to the actual server.

**Comment by chris-rl** [comment]
this would be very helpfull. I would like to use this to remove comments from the json request body. (because this isn't possible as well, see https://github.com/postmanlabs/postman-app-support/issues/3358 )

But I really don't like the workaround to put all the request data in the pre-request-script and just set the body from a variable. It is good to have this workaround, but it still is just a workaround.

**Comment by StrohmyerTR** [comment]
Linking this here, because it is a better description of the actual bug/enhancement (I do not agree with it being closed as a duplicate). https://github.com/postmanlabs/postman-app-support/issues/8466. Can someone please provide an update on this ticket. Thanks.

**Comment by jordanbisato** [comment]
2 years after the creation of issue and i'm here searching for the same thing. 
Any updates on this?

**Comment by StrohmyerTR** [comment]
IMO, this bug should be closed, and #8466 reopened. This thread does not give a concise description of what many of us are looking for. @codenirvana

**Comment by codenirvana** [comment]
Hello everyone 👋
**Postman v8.3.0** added support for mutating the request body in the `pre-request` scripts.

Example to set different body modes:
```js
// Raw
pm.request.body.update('Hello World!');

// Form Data
pm.request.body.update({
    mode: 'formdata',
    formdata: [{
        key: 'foo',
        value: 'bar'
    }]
});

// URL Encoded
pm.request.body.update({
    mode: 'urlencoded',
    urlencoded: [{
        key: 'foo',
        value: 'bar'
    }]
});

// GraphQL
pm.request.body.update({
    mode: 'graphql',
    graphql: {
        query: `
            query Square($ten: Int!) {
                square(n: $ten)
            }
        `,
        variables: { ten: 10 }
    }
});
```

Example to mutate existing request body:
```js
// Send Base64 encoded body
const rawBody = pm.request.body.toString();
const base64Body = btoa(rawBody);

pm.request.body.update(base64Body);
```

```js
// Add a new param
pm.request.body.formdata.add({
    key: 'foo',
    value: 'bar'
});
```

```js
// Strip JSON Comments
const rawData = pm.request.body.toString();
const strippedData = rawData.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m)

pm.request.body.update(strippedData);
```

Please try it out and let us know if you face any issues.

**Comment by lvl99** [comment]
Damn, need a micdrop emoji with that closed issue message 🎤 💥

**Comment by StrohmyerTR** [comment]
Hell yes. Thank you!!!

**Comment by jordanbisato** [comment]
A example to add json parameter on raw body: 

```
 var data = JSON.parse(pm.request.body.raw);
 data.key = key;
 let body = {
     mode: 'raw',
     raw: JSON.stringify(data),
     options: {
         raw: {
             language: 'json'
         }
     }
 }
 pm.request.body.update(body);
```

**Comment by tibi-extera** [comment]
Just yesterday I was looking into how to update the body in pre-request scripts. My example is getting data no older than 30 days. This would have to update every day, hence the need to update the request body.

Thank you for adding this feature. Works great. 😁👍🍾🥂

**Comment by vdespa** [comment]
> A example to add json parameter on raw body:
> 
> ```
>  var data = JSON.parse(pm.request.body.raw);
>     data.key = key;
> 
>     let body = {
>         mode: 'raw',
>         raw: JSON.stringify(data),
>         options: {
>             raw: {
>                 language: 'json'
>             }
>         }
>     }
>     pm.request.body.update(body);
> ```

For creating a request body from scratch, it makes sense to take this approach.

I think this can be simplified while **updating existing JSON**. Not sure if this is the intended usage, but the output _seems_ to be the same. 

    const body = JSON.parse(pm.request.body.raw);
    body.foo = 'bar';
    pm.request.body.raw = body;

Correct me if I am wrong @codenirvana

**Comment by flovogt** [comment]
Following code did it for me:
```
const rawData = pm.request.body.toString();
const changedData = rawData.replace("a", "b");
pm.request.body.update(changedData);
```

**Comment by coder-free** [comment]
@codenirvana Hi, is it possible to add a method to change the response too?
like this:
```
pm.response.body.update(changedData);
```

**Comment by shubhbhargav** [comment]
@coder-free Can you please provide a use-case where you need to update the response?

**Comment by coder-free** [comment]
@shubhbhargav 
like the request.
My server response data is encrypted:
```
{
    "status": 1,
    "msg": "success",
    "data": "g/sUMvvlDVyStLAE5B9KlGRP+m9EVbMfZb9fhTPNf8Ja ...... WmR8LaKQRM/OHF6vUbs+QDQCnpgIi"
}
```
Now my Tests script is:
```
var jsonBody = JSON.parse(responseBody);

if (jsonBody.data !== null && jsonBody.data !== undefined && jsonBody.data !== "null") {
    var decryptedData = MyDecryptFunc(jsonBody.data);
    jsonBody.data=decryptedData;
    console.log("jsonBody:"+ jsonBody);
    // responseBody = JSON.stringify(jsonBody);
    pm.response.body.update(JSON.stringify(jsonBody));
}
```
![image](https://user-images.githubusercontent.com/2630290/152170196-0b0b5772-a062-4da9-8959-886c7cd28562.png)

I just want to see the decrypted result conveniently.

This need is not strong. Because I can also see the result from the console log. It's just not that convenient.

Thanks.

**Comment by shubhbhargav** [comment]
This is interesting. Thanks for sharing it @coder-free. I would like to recommend an alternate to console (I know it's not ideal).
Have you tried the visualizer feature: https://learning.postman.com/docs/sending-requests/visualizer/? You can decrypt the response like the above and use pm.visualizer.set(...) to render it in the response tab under `Visualize`.

The idea of visualizer is to render customized responses, similar to this use-case. Let me know if that serves your use-case. :)

**Comment by StrohmyerTR** [comment]
You have two options - one, shubhbhargav, describes above. Your second option is to send your request in a pre-request script, parse the response, decrypt your string, and call pm.request.body.update() to be sent to the https://postman-echo.com/post API. I have an example of each, let me know if you would like to see an example of the second option I just described.

Here is an example of the first option, which will automatically copy your string to clipboard as well..:

`var isJson = pm.response.headers.all().some(header => {
  return header.value.includes('application/json');
});`

`var data = isJson ? MyDecryptFunc(pm.response.json().data) : null;`

`function MyDecryptFunc(string)
{
  return atob(string); // whatever resides in your decrypt function..
}`

`let template =
{{insert backtick here}}
<textarea id='copy' style='width:100%;height:98%;'></textarea>
<script>
  pm.getData(function (error, data)
  {
    var node = document.getElementById('copy');
    if (data.response === null)
    {
      node.textContent = 'Response was not in JSON format';
    }
    else
    {
      node.textContent = data.response;
      var selection = document.getSelection();
      selection.removeAllRanges();
      node.select();
      document.execCommand('copy');
      selection.removeAllRanges();
      document.getElementById('copy').textContent = "Data copied to clipboard:\\n\\n" + data.response;
    }
  });
</script>
{{insert backtick here}};`

`pm.visualizer.set(template, {
  response: data
});`

(Sorry for all the post edits.. Not sure how to write nested backticks in markup...)

**Comment by coder-free** [comment]
@shubhbhargav @StrohmyerTR Thanks. Visualizer is a great solution. Using it not only gives me the results I want. More additional features can be implemented if more time is invested.

**Comment by chris-rl** [comment]
@codenirvana 
> Hello everyone wave **Postman v8.3.0** added support for mutating the request body in the `pre-request` scripts.
> 
> (...)
> 
> ```js
> 
> // Form Data
> pm.request.body.update({
>     mode: 'formdata',
>     formdata: [{
>         key: 'foo',
>         value: 'bar'
>     }]
> });
> 
> ```
> (...)
> 
> Please try it out and let us know if you face any issues.

Hey!
Thanks for adding support to update the request body.

I'm struggling with setting the `file` in the form-data request.

My request looks like this:
```js
pm.sendRequest("https://upload.wikimedia.org/wikipedia/commons/thumb/3/35/Neckertal_20150527-6384.jpg/640px-Neckertal_20150527-6384.jpg", function (err, response) {
    const stream = response.stream;
    // const data = response.dataURI();
    
    // set Form Data request body
    pm.request.body.update({
        mode: 'formdata',
        formdata: [
            {
                key: 'id',
                value: pm.variables.replaceIn('{{$randomUUID}}')
            },
            {
                key: 'file',
                type: 'file',
                value: stream  // this is not working: "Not a valid image file"
                // value: data  // this is not working: "Not a valid image file"
            }
        ]
    });

});
```
How is it possible to set the file of a formdata using pre-request scripts?

**Comment by vdespa** [comment]
@chris-rl I guess the question here is how to submit a form-data request to upload a file without providing a file. AFAIK, this is not possible, not even from a regular Postman request.

I don't know a way to upload a file by directly providing the content.  Correct me if I am wrong.

**Comment by chris-rl** [comment]
@vdespa Yes, that's probably true. I'm not too much into the technical details on how form-data requests are working internally. But I guess at some point the _file_ will be loaded and sent to the endpoint, right? There should be a way of doing this manually.
Any ideas are appreciated.

Maybe another idea would be to save the downloaded file temporarily and then use the path to the temp file :thinking: 
Would this be possible in a pre-request script?

**Comment by vdespa** [comment]
@chris-rl I think the expectation is that the file exists somewhere on the disk. However, you can't write files from Postman, unless you rely on an externals script that does that for you.

**Comment by chris-rl** [comment]
@vdespa Hm, how would using an external script work?

Another idea: Would it be possible to point to a file that is not located on a local disk, but in a remote location? Maybe just provide the URL where it could be downloaded?

Anyways, thanks for your input @vdespa!
